### PR TITLE
test(aio): add async beforeEach to prevent Chrome disconnects

### PR DESCRIPTION
**Pull request by petebacondarwin**
Related to 3d5f520ff08f00f89f323f1ec3bcd1e0c3f38596 from #17405

### CODE REVIEW

**Comment by petebacondarwin** [comment]
Here is a capture of a unit test run that disconnects:

image


The disconnection was happening between 54 and 64 seconds:

```
LOG: 179
Chrome 59.0.3071 (Mac OS X 10.12.5): Executed 179 of 485 (skipped 2) SUCCESS (0 secs / 54.58 secs)
LOG: 179
Chrome 59.0.3071 (Mac OS X 10.12.5) ERROR
  Disconnected, because no message in 10000 ms.
Chrome 59.0.3071 (Mac OS X 10.12.5): Executed 179 of 485 (skipped 2) DISCONNECTED (1 min 4.853 secs / 54.58 secs)
Chrome 59.0.3071 (Mac OS X 10.12.5) ERROR
```

You can see that the browser is happily running the JS tests still during this period

What I believe is happening is that Chrome 59 has some new optimization that blocks the rendering thread (and also the thread that may be used to push data to the socket on which Karma is listening) when the JS gets too intense?

This might explain, perhaps, why people are able to work around this by putting async tests (with `setTimeout(noop, 0)`) into their specs. The async bits would cause the browser to exit its intense JS operation and allow the socket to be accessed

Here is a capture with the async beforeEach being run on every test:

image

You can see that during in the similar period that was causing the disconnect before, the memory and node usage is the same. But now we are getting much greater frame rates (~15fps), which is probably because our async beforeEach is freeing the browser to do rendering and also to talk to the socket.

File: aio/src/test.ts [hunk 1]
```diff
@@ -42,6 +42,14 @@ getTestBed().initTestEnvironment(
   platformBrowserDynamicTesting()
 );
 
+// HACK: Chrome 59 disconnects if there are too many synchronous tests in a row
+// because it appears to lock up the thread that communicates to Karma's socket
+// This async beforeEach gets called on every spec and releases the JS thread long
+// enough for the socket to continue to communicate.
+// The downside is that it creates a minor performance penalty of around 10-15%
+// increase in the time it takes to run out unit tests.
+beforeEach((done) => done());
```
Comment (gkalpak) [inline aio/src/test.ts]: Interesting. I thought you can only have `beforeEach` inside `describe` blocks.

**Review by gkalpak — APPROVED** [review]
LGTM as long as it makes Travis happy.

**Comment by dauledk** [comment]
@mary-poppins broken link

**Comment by gkalpak** [comment]
@dauledk, the preview links posted by @mary-poppins are only valid for as long as the PR is open (they are only for preview purposes). That said, there is nothing in the PR that would affect the app.

**Pull request merged**